### PR TITLE
Add Philips Hue hub and lightbulb.

### PR DESCRIPTION
**Pull request by m-chau**
Adding Hue helper library and updated the Hue bulb translator to reflect the OCF schema.

### CODE REVIEW

File: org.opent2t.sample.lamp.superpopular/com.philips.hue.bulb/js/thingTranslator.js [hunk 1]
```diff
@@ -40,56 +16,133 @@ function validateArgumentType(arg, argName, expectedType) {
...
+}
+
+var deviceId;
+var deviceType = 'lights';
+var hueHelper;
+
+// This translator class implements the 'org.opent2t.sample.light.superpopular' schema.
+class Translator {
+
+    constructor(device) {
```
Comment (danielrischar) [inline org.opent2t.sample.lamp.superpopular/com.philips.hue.bulb/js/thingTranslator.js]: The Nest and Wink translators have moved away from having *helper classes, and instead use a concept of "hubs" even where a physical hub doesn't exist (like with Nest).  The "hub" is then passed to the translator in the constructor.

File: Helpers/opent2t-translator-helper-hue/js/index.js [hunk 2]
```diff
@@ -0,0 +1,127 @@
+'use strict';
```
Comment (danielrischar) [inline Helpers/opent2t-translator-helper-hue/js/index.js]: This should follow the convention of "hubs" as "providers" that the Nest and Wink translators currently use.  PRobably a com.hue.hub translator that can then be used by the device translators.  See Arjun's recent changes for Nest as an example.  Nest, like Hue, doesn't have a physical hub, so in this case the hub is a representation of a single API/service that is used to talk to the devices.

File: Helpers/opent2t-translator-helper-hue/js/index.js [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+class HueHelper {
+
+    // Init Helper: Sets the initial parameters to build our target endpoint
+    constructor(accessToken, bridgeId, whitelistId) {
+        bearerToken = accessToken;
+		apiEndpoint =  apiBaseAddress + bridgeId + '/' + whitelistId;
+        console.log('Initialized Hue Helper.');
+    }
+
+    // Gets device details (all fields), response formatted per http://docs.winkapiv2.apiary.io/
```
Comment (danielrischar) [inline Helpers/opent2t-translator-helper-hue/js/index.js]: Should be http://www.developers.meethue.com/philips-hue-api

File: Helpers/opent2t-translator-helper-hue/js/tests/README.md [hunk 4]
```diff
@@ -0,0 +1,24 @@
+# Running Tests
```
Comment (arjun-msft) [inline Helpers/opent2t-translator-helper-hue/js/tests/README.md]: This readme should be in the /js folder and should have the readme for the helper (hub after you move it) and tests combined

File: Helpers/opent2t-translator-helper-hue/js/index.js (hunk 2 above)
Comment (arjun-msft) [inline Helpers/opent2t-translator-helper-hue/js/index.js]: Agreed, you should have two examples (nest hub and wink hub) to follow.

---
In reply to: [78867509](https://github.com/openT2T/translators/pull/96#discussion_r78867509) [](ancestors = 78867509)

File: Helpers/opent2t-translator-helper-hue/js/tests/test.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+putPayload['bri'] = 20;
+putPayload['name'] = "Hue Color Lamp";
+
+return helper.putDeviceDetailsAsync(testConfig.deviceType, testConfig.deviceId, putPayload)
+    .then(result => {
+        console.log('set multiple result: ' + JSON.stringify(result, null, 2));
+
+        // get all fields at the same time
+        return helper.getDeviceDetailsAsync(testConfig.deviceType, testConfig.deviceId)
+            .then(result2 => {
```
Comment (arjun-msft) [inline Helpers/opent2t-translator-helper-hue/js/tests/test.js]: >result2 [](start = 18, length = 7)

you should check the resutlt2 also.

File: org.opent2t.sample.lamp.superpopular/com.philips.hue.bulb/js/thingTranslator.js [hunk 6]
```diff
@@ -40,56 +16,133 @@ function validateArgumentType(arg, argName, expectedType) {
     }
 }
 
+// Helper method to convert the device schema to the translator schema.
+function deviceSchemaToTranslatorSchema(deviceSchema) {
+
+    var deviceState = deviceSchema.state;
+    var result = {
+        id:  'lights.' + deviceId,
```
Comment (arjun-msft) [inline org.opent2t.sample.lamp.superpopular/com.philips.hue.bulb/js/thingTranslator.js]: >'lights.' +  [](start = 13, length = 12)

lets keep this just deviceId. the others need to be updated to just deviceId

File: org.opent2t.sample.lamp.superpopular/com.philips.hue.bulb/js/tests/test.js [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+        .then(translator => {
+            // TEST: translator is valid
+            t.is(typeof translator, 'object') && t.truthy(translator);
+
+            // set value to true
+            return OpenT2T.setPropertyAsync(translator, 'org.opent2t.sample.lamp.superpopular', 'power', true)
+                .then(() => {
+
+                    // wait a bit...
+                    return sleep(20000).then(() => {
```
Comment (arjun-msft) [inline org.opent2t.sample.lamp.superpopular/com.philips.hue.bulb/js/tests/test.js]: >20000 [](start = 33, length = 5)

nit: seems high

Comment (m-chau) [inline org.opent2t.sample.lamp.superpopular/com.philips.hue.bulb/js/tests/test.js]: It's because Hut takes over 15 sec to complete a remote request.

File: Helpers/opent2t-translator-helper-hue/js/index.js (hunk 2 above)
Comment (m-chau) [inline Helpers/opent2t-translator-helper-hue/js/index.js]: Migrated to using Hue hub.

File: Helpers/opent2t-translator-helper-hue/js/index.js (hunk 3 above)
Comment (m-chau) [inline Helpers/opent2t-translator-helper-hue/js/index.js]: Updated.

File: org.opent2t.sample.lamp.superpopular/com.philips.hue.bulb/js/thingTranslator.js (hunk 1 above)
Comment (m-chau) [inline org.opent2t.sample.lamp.superpopular/com.philips.hue.bulb/js/thingTranslator.js]: Migrated to using Hue hub.

File: org.opent2t.sample.lamp.superpopular/com.philips.hue.bulb/js/thingTranslator.js (hunk 6 above)
Comment (m-chau) [inline org.opent2t.sample.lamp.superpopular/com.philips.hue.bulb/js/thingTranslator.js]: Fixed in the latest commit.

File: org.opent2t.sample.hub.superpopular/com.hue.hub/js/tests/test.js [hunk 8]
```diff
@@ -0,0 +1,41 @@
+var test = require('ava');
+var OpenT2T = require('opent2t').OpenT2T;
+var config = require('./testConfig');
+
+// var schema = require("c:/githome/translators/org.opent2t.sample.thermostat.superpopular/org.opent2t.sample.thermostat.superpopular");
+// console.log(schema);
+
+var test = require('ava');
+var OpenT2T = require('opent2t').OpenT2T;
+var config = require('./testConfig');
```
Comment (begillil) [inline org.opent2t.sample.hub.superpopular/com.hue.hub/js/tests/test.js]: This is all declared twice.

File: org.opent2t.sample.hub.superpopular/com.hue.hub/js/tests/test.js [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+// var schema = require("c:/githome/translators/org.opent2t.sample.thermostat.superpopular/org.opent2t.sample.thermostat.superpopular");
+// console.log(schema);
```
Comment (begillil) [inline org.opent2t.sample.hub.superpopular/com.hue.hub/js/tests/test.js]: Should remove this.

File: org.opent2t.sample.lamp.superpopular/com.philips.hue.bulb/js/tests/test.js [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+
+var translatorPath = require('path').join(__dirname, '..');
+var hubPath = require('path').join(__dirname, '../../../../org.opent2t.sample.hub.superpopular/com.hue.hub/js');
+var translator = undefined;
+
+function getLamp(devices) {
+    for (var i = 0; i < devices.length; i++) {
+        var d = devices[i];
+
+        if (d.openT2T.translator === 'opent2t-translator-com-hue-bulb' && d.id === '7') {
```
Comment (begillil) [inline org.opent2t.sample.lamp.superpopular/com.philips.hue.bulb/js/tests/test.js]: >&& d.id === '7' [](start = 71, length = 15)

Will every account have a bulb with an id of 7?

File: Helpers/opent2t-translator-helper-hue/js/package.json [hunk 11]
```diff
@@ -0,0 +1,12 @@
+{
+    "name": "opent2t-translator-helper-hue",
```
Comment (begillil) [inline Helpers/opent2t-translator-helper-hue/js/package.json]: >"opent2t-translator-helper-hue", [](start = 12, length = 32)

Do we still need this helper now that you've moved to the 'hub' model?

File: org.opent2t.sample.hub.superpopular/com.hue.hub/js/tests/test.js (hunk 8 above)
Comment (m-chau) [inline org.opent2t.sample.hub.superpopular/com.hue.hub/js/tests/test.js]: Fixed

File: org.opent2t.sample.hub.superpopular/com.hue.hub/js/tests/test.js (hunk 9 above)
Comment (m-chau) [inline org.opent2t.sample.hub.superpopular/com.hue.hub/js/tests/test.js]: Fixed.

File: org.opent2t.sample.lamp.superpopular/com.philips.hue.bulb/js/tests/test.js (hunk 10 above)
Comment (m-chau) [inline org.opent2t.sample.lamp.superpopular/com.philips.hue.bulb/js/tests/test.js]: Fixed. That was added for my own testing purpose.

File: Helpers/opent2t-translator-helper-hue/js/package.json (hunk 11 above)
Comment (m-chau) [inline Helpers/opent2t-translator-helper-hue/js/package.json]: I do not think that we need it. Since the module was completed, I thought I would just add it anyway. Any suggest if we should add or remove the hue helper module?

File: org.OpenT2T.Sample.SuperPopular.Thermostat/Wink Thermostat/js/test.js [hunk 12]
```diff
@@ -0,0 +1,39 @@
+var translator = require('./thingTranslator');
```
Comment (begillil) [inline org.OpenT2T.Sample.SuperPopular.Thermostat/Wink Thermostat/js/test.js]: It looks like somehow this file was accidentally added to the PR.

File: Helpers/opent2t-translator-helper-hue/js/index.js [hunk 13]
```diff
@@ -0,0 +1,127 @@
+'use strict';
+
+var request = require('request-promise');
```
Comment (arjun-msft) [inline Helpers/opent2t-translator-helper-hue/js/index.js]: This file should be removed right?

File: org.opent2t.sample.hub.superpopular/com.hue.hub/js/package.json [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+    "url": "githttps://github.com/openT2T/translators.git"
+  },
+  "bugs": {
+    "url": "https://github.com/openT2T/translators/issues"
+  },
+  "homepage": "https://github.com/openT2T/translators#readme",
+  "dependencies": {
+    "request": "^2.74.0",
+    "request-promise": "^4.1.1",
+    "opent2t": "^1.0.1"
```
Comment (arjun-msft) [inline org.opent2t.sample.hub.superpopular/com.hue.hub/js/package.json]: should have a reference to the hue onboarding

File: org.opent2t.sample.hub.superpopular/com.hue.hub/js/manifest.xml [hunk 15]
```diff
@@ -0,0 +1,37 @@
+<?xml version="1.0" encoding="utf-8"?>
+<manifest>
```
Comment (arjun-msft) [inline org.opent2t.sample.hub.superpopular/com.hue.hub/js/manifest.xml]: hue needs the exact same info as wink?

Comment (arjun-msft) [inline org.opent2t.sample.hub.superpopular/com.hue.hub/js/manifest.xml]: Looks like this should be the same as nest not wink.  based on the hue onboarding PR

---
In reply to: [80634569](https://github.com/openT2T/translators/pull/96#discussion_r80634569) [](ancestors = 80634569)

File: org.opent2t.sample.lamp.superpopular/com.philips.hue.bulb/js/package.json [hunk 16]
```diff
@@ -1,21 +1,31 @@
...
+    },
+    "ava": {
+        "files": [
+            "tests/*.js"
+        ]
+    },
+    "dependencies": {
+      "opent2t-translator-com-hue-hub": "*",
+      "opent2t": "*",
+      "opent2t-onboarding-huehub": "*"
```
Comment (arjun-msft) [inline org.opent2t.sample.lamp.superpopular/com.philips.hue.bulb/js/package.json]: remove and put in the hub.

**Comment by arjun-msft** [comment]
:clock1:

File: org.opent2t.sample.lamp.superpopular/com.philips.hue.bulb/js/package.json (hunk 16 above)
Comment (m-chau) [inline org.opent2t.sample.lamp.superpopular/com.philips.hue.bulb/js/package.json]: Fixed according to wink lamp's package.json[https://github.com/openT2T/translators/blob/dev/org.opent2t.sample.lamp.superpopular/com.wink.lightbulb/js/package.json].

File: org.opent2t.sample.hub.superpopular/com.hue.hub/js/manifest.xml (hunk 15 above)
Comment (m-chau) [inline org.opent2t.sample.hub.superpopular/com.hue.hub/js/manifest.xml]: Fixed.

File: org.opent2t.sample.hub.superpopular/com.hue.hub/js/package.json (hunk 14 above)
Comment (m-chau) [inline org.opent2t.sample.hub.superpopular/com.hue.hub/js/package.json]: Added.

File: Helpers/opent2t-translator-helper-hue/js/index.js (hunk 13 above)
Comment (m-chau) [inline Helpers/opent2t-translator-helper-hue/js/index.js]: @arjun-msft  Should I remove the Hue helper library all together?

File: org.OpenT2T.Sample.SuperPopular.Thermostat/Wink Thermostat/js/test.js (hunk 12 above)
Comment (m-chau) [inline org.OpenT2T.Sample.SuperPopular.Thermostat/Wink Thermostat/js/test.js]: removed from PR.

File: org.opent2t.sample.hub.superpopular/com.hue.hub/js/common.js [hunk 17]
```diff
@@ -0,0 +1,22 @@
+/* jshint esversion: 6 */
+/* jshint node: true */
```
Comment (ChuckFerring) [inline org.opent2t.sample.hub.superpopular/com.hue.hub/js/common.js]: Needed?

File: Helpers/opent2t-translator-helper-hue/js/tests/README.md (hunk 4 above)
Comment (m-chau) [inline Helpers/opent2t-translator-helper-hue/js/tests/README.md]: Both wink and nest helpers have the readme under /js/tests and did not have the readme for the helper itself. Should we keep hue helper consistent with wink and nest helpers?

File: org.opent2t.sample.hub.superpopular/com.hue.hub/js/common.js (hunk 17 above)
Comment (m-chau) [inline org.opent2t.sample.hub.superpopular/com.hue.hub/js/common.js]: Since both wink and nest included this, I included it in hue for consistency. @arjun-msft any comments?

File: Helpers/opent2t-translator-helper-hue/js/package.json (hunk 11 above)
Comment (arjun-msft) [inline Helpers/opent2t-translator-helper-hue/js/package.json]: Since this is a new helper module, I would suggest not adding it.

---
In reply to: [80416298](https://github.com/openT2T/translators/pull/96#discussion_r80416298) [](ancestors = 80416298)

File: Helpers/opent2t-translator-helper-hue/js/tests/README.md (hunk 4 above)
Comment (arjun-msft) [inline Helpers/opent2t-translator-helper-hue/js/tests/README.md]: since this is going away, don't worry about the readme

---
In reply to: [81488209](https://github.com/openT2T/translators/pull/96#discussion_r81488209) [](ancestors = 81488209)

File: org.opent2t.sample.hub.superpopular/com.hue.hub/js/common.js (hunk 17 above)
Comment (arjun-msft) [inline org.opent2t.sample.hub.superpopular/com.hue.hub/js/common.js]: I'm not sure at the moment, this has been all over the place.  My recommendation is to leave it to be consistent and we can remove it later for everyone if thats the way we want to go.

---
In reply to: [81489009](https://github.com/openT2T/translators/pull/96#discussion_r81489009) [](ancestors = 81489009)

File: org.opent2t.sample.thermostat.superpopular/com.wink.thermostat/js/tests/test.js [hunk 18]
```diff
@@ -1,293 +1,195 @@
...
-    // Once the notification has been received, unsubscribe and end the test
-    return deferred.promise.then(() => {
-        console.log("Unsubscribing");
-        return translator.deleteSubscribeThermostatResURI(callbackUrlParams).then(() => {
-            t.pass("Unsubscribed successfully");
-        });
-    });
-});
+var test = require('ava');
+var OpenT2T = require('opent2t').OpenT2T;
```
Comment (arjun-msft) [inline org.opent2t.sample.thermostat.superpopular/com.wink.thermostat/js/tests/test.js]: are these changes needed for your pull request?

File: Helpers/opent2t-translator-helper-hue/js/index.js (hunk 13 above)
Comment (arjun-msft) [inline Helpers/opent2t-translator-helper-hue/js/index.js]: Since you are adding the helper, I'd say don't add it since we are now using the hub.

---
In reply to: [81057804](https://github.com/openT2T/translators/pull/96#discussion_r81057804) [](ancestors = 81057804)

File: org.opent2t.sample.hub.superpopular/com.hue.hub/js/manifest.xml [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+      <flow>
+        <arg name="client_secret" />
+        <description language="en">Ask for Client Secret</description>
+        <description language="fr">Demandez Client Secret</description>
+      </flow>
+    </onboardflow>
+    <onboardflow name="askUserPermission">
+      <flow>
+        <arg name="url" />
+        <description language="en">https://api.meethue.com/oauth2/auth?clientid={client_id}&appid={app_id}&deviceid={device_id}&devicename={device_name}&state={state}&response_type=code</description>
```
Comment (arjun-msft) [inline org.opent2t.sample.hub.superpopular/com.hue.hub/js/manifest.xml]: the & need to be changed to &amp;

without the &amp; it fails syntax test, you can verify using notepad++ and xml tools, verify xml syntax

File: org.opent2t.sample.lamp.superpopular/com.philips.hue.bulb/js/package.json [hunk 20]
```diff
@@ -1,21 +1,31 @@
...
-    "node-hue-api": "~2.2.0"
-  },
-  "devDependencies": {
-    "optimist": "~0.6.1",
-    "readline": "~1.3.0",
-    "opent2t-onboarding-huehub": "*"
-  },
-  "author": "",
-  "license": "MIT"
+    "name": "opent2t-translator-com-hue-bulb",
```
Comment (arjun-msft) [inline org.opent2t.sample.lamp.superpopular/com.philips.hue.bulb/js/package.json]: the name of the package is correct, can you update the folder name to align with this?  instead of com.philips.hue.bulb it should be com.hue.bulb

File: org.opent2t.sample.hub.superpopular/com.hue.hub/js/manifest.xml [hunk 21]
```diff
@@ -0,0 +1,29 @@
...
+        <arg name="client_id" />
+        <description language="en">Ask for Client ID</description>
+        <description language="fr">Demandez Client ID</description>
+      </flow>
+      <flow>
+        <arg name="client_secret" />
+        <description language="en">Ask for Client Secret</description>
+        <description language="fr">Demandez Client Secret</description>
+      </flow>
+    </onboardflow>
```
Comment (arjun-msft) [inline org.opent2t.sample.hub.superpopular/com.hue.hub/js/manifest.xml]: add app_id, device_id and device_name to this section

File: org.opent2t.sample.hub.superpopular/com.hue.hub/js/manifest.xml [hunk 22]
```diff
@@ -0,0 +1,29 @@
+<?xml version="1.0" encoding="utf-8"?>
+<manifest>
+  <schemas>
+      <schema id="org.opent2t.sample.hub.superpopular" main="true" type="OCF" />
+      <schema id="oic.core" type="OCF" />
+      <schema id="oic.baseResource" type="OCF" />
+  </schemas>
+  <!-- associated onboarding module -->
+  <onboarding id="org.opent2t.onboarding.huehub">
```
Comment (arjun-msft) [inline org.opent2t.sample.hub.superpopular/com.hue.hub/js/manifest.xml]: this should be the same name as the onboarding package, in this case org.opent2t.onboarding.hue

Comment (m-chau) [inline org.opent2t.sample.hub.superpopular/com.hue.hub/js/manifest.xml]: Fixed.

File: org.opent2t.sample.hub.superpopular/com.hue.hub/js/manifest.xml (hunk 21 above)
Comment (m-chau) [inline org.opent2t.sample.hub.superpopular/com.hue.hub/js/manifest.xml]: Fixed.

File: org.opent2t.sample.hub.superpopular/com.hue.hub/js/manifest.xml (hunk 19 above)
Comment (m-chau) [inline org.opent2t.sample.hub.superpopular/com.hue.hub/js/manifest.xml]: Fixed.
